kubo-migrator: add migration from 12 to 13

Luflosi opened the pull request:
###### Description of changes
https://github.com/ipfs/fs-repo-migrations/releases/tag/fs-repo-12-to-13%2Fv1.0.0
This will be used to upgrade the Kubo repo when updating to Kubo 0.18.

Two old migrations were broken since Go 1.16 was removed from Nixpkgs.
Upstream is not interested in fixing the migrations to work with newer Go versions: https://github.com/ipfs/fs-repo-migrations/pull/163#issuecomment-1397311134.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Luflosi committed:
kubo-migrator: add migration from 12 to 13

https://github.com/ipfs/fs-repo-migrations/releases/tag/fs-repo-12-to-13%2Fv1.0.0
This will be used to upgrade the Kubo repo when updating to Kubo 0.18.

milahu commented:
> Replace them with a stub as discussed in [#168480 (comment)](https://github.com/NixOS/nixpkgs/pull/168480#issuecomment-1118398642), so that `fs-repo-migrations` still finds the binaries in the PATH and never attempts to download these binaries from the internet, because that is a bit uncool.

this fails because [ld-linux.so is missing](https://github.com/ipfs/kubo/issues/2907#issuecomment-1378801936)

> If anyone still wants to actually use these migrations, they should use an old version of Nixpkgs that still includes Go 1.16.

not practical, as this will compile go_1_16 = slow

the fix is as simple as https://github.com/ipfs/fs-repo-migrations/pull/163

Luflosi commented:
> this fails because ld-linux.so is missing

From reading what you linked to, it sounds like you ran `ipfs daemon` with a repo, which required a migration. The migration binary `fs-repo-11-to-12` was not in the `$PATH`, so Kubo downloaded it off the internet, which then failed to run. It doesn't surprise me at all that random binaries off of the internet don't work on NixOS because NixOS does not implement the FHS. As such I don't think the downloaded binary not working on NixOS is an issue in Kubo. Please make that clear in the linked issue.
Kubo would not have tried to download a binary if you had put `kubo-migrator-all-fs-repo-migrations` into its `$PATH`. I replaced all the old migrations instead of deleting them to try to prevent such a situation.
Anyways, my comment only mentions `fs-repo-migrations`, which you did not run and so I don't understand what your comment is about. Did I understand something incorrectly?

> not practical, as this will compile go_1_16 = slow

This is not true, everything is still in the cache. To test for yourself:
```sh
git checkout 4c3c80df545ec5cb26b5480979c3e3f93518cbe5
nix-build -A ipfs-migrator
```
I got this commit hash from the list at https://channels.nix.gsc.io/nixpkgs-unstable/history. The timestamp is from a day before Go 1.16 was removed.

> the fix is as simple as ipfs/fs-repo-migrations#163

Thanks, I will try to incorporate this patch into this PR when I have some time.
You only fixed the 10-to-11 and 11-to-12 migrations. What about the older migrations? Can they not be fixed in the same way or would it be more effort?

Luflosi committed:
kubo-migrator: fix broken migrations

Two old migrations were broken since Go 1.16 was removed from Nixpkgs.
Upstream is not interested in fixing the migrations to work with newer Go versions: https://github.com/ipfs/fs-repo-migrations/pull/163#issuecomment-1397311134.

Luflosi commented:
Ok, it should work now, thanks again for the patch.
I had to create a separate derivation for the source. This is because I'm using using `sourceRoot` inside the derivation which builds the individual migrations and that would otherwise require me to change the patches to apply to the child directories instead and to apply each patch only to the corresponding migration. Using a separate derivation just for patching is easier.

milahu commented on pkgs/applications/networking/kubo-migrator/src.nix:
```diff
@@ -0,0 +1,42 @@
+    hash = "sha256-QQone7E2Be+jVfnrwqQ1Ny4jo6mSDHhaY3ErkNdn2f8=";
+  };
+
+  patches = [
+    # Make the migrations compatible with Go 1.17 and newer
+    (fetchpatch {
+      name = "fix-fs-repo-10-to-11.patch";
+      url = "https://github.com/ipfs/fs-repo-migrations/pull/163/commits/c3b39a3ccd51e65e1039652235cb6930d3304a08.patch";
+      hash = "sha256-YkFtFsYueCd24WsRINq7Q5wvFjoIVAqlN9yy9y6kNrw=";
+    })
+    (fetchpatch {
+      name = "fix-fs-repo-11-to-12.patch";
+      url = "https://github.com/ipfs/fs-repo-migrations/pull/163/commits/408f58ffdf9869bcb51cd4d8d99d5aefd65df593.patch";
+      hash = "sha256-IkRkwNVDLSU8wWmtFuxM5H0g6Jkkso/i8k5XmjMrLA8=";
+    })
+  ];
```
i suggest to commit this as one patchfile to nixpkgs

Luflosi replied:
Why? Your PR is not going to disappear. How would committing the patch to Nixpkgs be better?

milahu replied:
generally, we should avoid dependency on github
the source can be hosted on other git providers, and then the patch is missing

Luflosi replied:
I have archived the patches: https://web.archive.org/web/20230125112511/https://github.com/ipfs/fs-repo-migrations/commit/c3b39a3ccd51e65e1039652235cb6930d3304a08.patch https://web.archive.org/web/20230125112541/https://github.com/ipfs/fs-repo-migrations/commit/408f58ffdf9869bcb51cd4d8d99d5aefd65df593.patch. If the source ever disappears for any reason, we can still access the patches from archive.org.
I try to avoid adding patches to Nixpkgs directly unless necessary to keep the size down. I think fetching a patch like this is perfectly acceptable in this case.

milahu commented on pkgs/applications/networking/kubo-migrator/unwrapped.nix:
```diff
@@ -1,27 +1,17 @@
-    rev = "fs-repo-11-to-12/v1.0.2";
-    sha256 = "sha256-CG4utwH+/+Igw+SP3imhl39wijlB53UGtkJG5Mwh+Ik=";
-  };
+  src = callPackage ./src.nix { };
```
why so complicated?

Luflosi replied:
What else do you suggest?

milahu commented on pkgs/applications/networking/kubo-migrator/src.nix:
```diff
@@ -0,0 +1,42 @@
+stdenvNoCC.mkDerivation rec {
+  name = "kubo-migrator-src";
+
+  src = fetchFromGitHub {
+    owner = "ipfs";
+    repo = "fs-repo-migrations";
+    rev = "fs-repo-12-to-13/v1.0.0";
+    hash = "sha256-QQone7E2Be+jVfnrwqQ1Ny4jo6mSDHhaY3ErkNdn2f8=";
+  };
+
+  patches = [
+    # Make the migrations compatible with Go 1.17 and newer
+    (fetchpatch {
+      name = "fix-fs-repo-10-to-11.patch";
+      url = "https://github.com/ipfs/fs-repo-migrations/pull/163/commits/c3b39a3ccd51e65e1039652235cb6930d3304a08.patch";
+      hash = "sha256-YkFtFsYueCd24WsRINq7Q5wvFjoIVAqlN9yy9y6kNrw=";
+    })
+    (fetchpatch {
+      name = "fix-fs-repo-11-to-12.patch";
+      url = "https://github.com/ipfs/fs-repo-migrations/pull/163/commits/408f58ffdf9869bcb51cd4d8d99d5aefd65df593.patch";
+      hash = "sha256-IkRkwNVDLSU8wWmtFuxM5H0g6Jkkso/i8k5XmjMrLA8=";
+    })
+  ];
```
move all this into `buildGoModule rec { ... }`

Luflosi replied:
You mean in `pkgs/applications/networking/kubo-migrator/unwrapped.nix`? I tried that but it doesn't work because the individual migrations use `kubo-migrator-unwrapped.src` to get the source code and that will not be patched.

milahu commented on pkgs/applications/networking/kubo-migrator/all-migrations.nix:
```diff
@@ -23,8 +23,9 @@ let
   };
```
use overrideAttrs to inherit all attributes, including `src` and `patches`

```nix
{
  fs-repo-common = pname: version: kubo-migrator-unwrapped.overrideAttrs (old: {
    inherit pname version;
    sourceRoot = "kubo-migrator-src/${pname}";
    vendorSha256 = null;
    doCheck = false;
    meta = old.meta // {
      mainProgram = pname;
      description = "Individual migration for the filesystem repository of Kubo clients";
    };
  });

```

Luflosi replied:
This doesn't work either because of `sourceRoot` as I already wrote in https://github.com/NixOS/nixpkgs/pull/209304#issuecomment-1402559385. Please try it yourself.

milahu replied:
> Please try it yourself.

good idea ^^

> go: inconsistent vendoring in /build/source/fs-repo-12-to-13

https://discourse.nixos.org/t/buildgomodule-with-local-src-inconsistent-vendoring/8641

> It looks like this is how a bad vendorSha256 manifests itself

https://discourse.nixos.org/t/inconsistent-vendoring-in-buildgomodule-when-overriding-source/9225/2

> overrideAttrs and buildGoModule aren’t aware of each other

https://github.com/NixOS/nixpkgs/issues/86349

&rarr; override buildGoModule
https://github.com/milahu/nixpkgs/tree/Luflosi/update/kubo-migrator

the problem is that overrideAttrs does not work, so `vendorHash != null`, so its using the wrong vendor folder

configurePhase:

```sh
runHook preConfigure

export GOCACHE=$TMPDIR/go-cache
export GOPATH="$TMPDIR/go"
export GOPROXY=off
export GOSUMDB=off
cd "$modRoot"
rm -rf vendor
cp -r --reflink=auto /nix/store/nd7scmc52v6bm2qvh7vwdqcd74a9bjzw-kubo-migrator-2.0.2-go-modules vendor


runHook postConfigure
```

see also
pkgs/build-support/go/module.nix

Luflosi replied:
This looks MUCH more complicated than what I'm doing, for seemingly no gain. Do you agree? Should this PR be merged as-is?

milahu replied:
looks complicated, because `buildGoModule.overrideGoAttrs` is missing in nixpkgs (also because i added lots of noise in my comments and drafts...)

but "usually" in nixpkgs, we create one base derivation, and from there, create multiple derivations using overrideAttrs

creating intermediary derivations (like youre doing) i would call an anti-pattern, because its a waste of disk space... but at the end of the day, it should "just work", so feel free to merge ; )

forgondolin replied:
Don't want to stop the idea here, but if I remember, the biggest problem in the migrator was that [qlts](https://github.com/ipfs/fs-repo-migrations/issues/157)
Wouldn't be easier try to upgrade this specific package to 1.18? Just for curiosity.

milahu replied:
> Wouldn't be easier try to upgrade this specific package to 1.18?

done in `patches`

milahu replied:
just read my reviews ...

my patches from https://github.com/ipfs/fs-repo-migrations/pull/163 can be reduced to something like

```nix
{
  # fix build on go 1.18: panic: qtls.ClientHelloInfo doesn't match
  # https://github.com/ipfs/fs-repo-migrations/pull/163
  patchPhase = ''
    substituteInPlace vendor/github.com/marten-seemann/qtls-go1-15/common.go \
      --replace \
        '"container/list"' \
        '"container/list"
        "context"' \
      --replace \
        'config *Config' \
        'config *Config
        ctx context.Context'
  '';
```

(i hope that go wont cry about the broken indents)

if you have more questions you can ask me on matrix `@milahu:matrix.org`

forgondolin replied:
Right, that's seems reasonable. Currently I'm using `steam-run` or `nix-alien` manually to fix this on my nodes, I'm thinking on using your branch for it (at least for now, because running this by hand is not a good practice).

Luflosi replied:
I updated this PR to use `substituteInPlace` as suggested above. This is much cleaner.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/850
